File: 101/Graph/Connected-Components/Max-Area-of-Island/solution.py
#https://leetcode.com/problems/max-area-of-island/
class Solution(object):
    dirns = [
        [-1,0],
        [1,0],
        [0,1],
        [0,-1]
    ]
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        if not grid :
            return 0
        rowSz = len(grid)
        colSz = len(grid[0])
        result = 0
        for r in range(0,rowSz):
            for c in range(0, colSz):
                if grid[r][c] == 0:
                    continue
                sz = self.dfs(grid, r,c)
                print(sz)
                result = max(result,sz)
        
        return result

        
    def dfs(self, grid, r, c):
        # we laready make sure that it is a land
        # visit it and mark it visited so that we dont visit it again
        sz = 1
        grid[r][c] = 0
        # try visiting nbrs
        rowSz = len(grid)
        colSz = len(grid[0])
        for dirn in self.dirns:
            nr = r + dirn[0]
            nc = c + dirn[1]
            if nr < 0 or nr >= rowSz or nc < 0 or nc >= colSz or grid[nr][nc] == 0:
                continue
            # found land
            sz+= self.dfs(grid, nr,nc)
        return sz

File: 101/Graph/Connected-Components/Max-Area-of-Island/solution.cs
public class Solution {
    public int MaxAreaOfIsland(int[][] grid) {
        int rowSz = grid.Length, colSz = grid[0].Length;
        int result = 0;
        bool[,] visited = new bool[rowSz, colSz];
        for(int r = 0; r < rowSz; ++r) {
            for(int c = 0; c < colSz; ++c) {
                if(grid[r][c] == 0 || visited[r,c] == true) {
                    continue;
                }
                int area = Explore(grid, r,c, visited);
                result = Math.Max(area, result);
            }
        }
        return result;
    }

    private int[,] dirns = new int[,] { {-1,0}, {1,0}, {0, 1}, {0, -1}};

    private int Explore(int[][] grid, int row, int col, bool[,] visited) {
        int rowSz = grid.Length, colSz = grid[0].Length;
        int area = 1;
        visited[row,col] = true;
        for(int d = 0; d <dirns.GetLength(0); ++d) {
            int nr = row + dirns[d,0];
            int nc = col + dirns[d,1];
            if(nr < 0 || nr >= rowSz || nc < 0 || nc >= colSz || 
                grid[nr][nc] == 0 || visited[nr,nc]) {
                    continue;
            }
            area += Explore(grid, nr, nc, visited);
        }
        return area;
    }
}